### ISSUE TITLE
sgx-driver/sgx-gdb: Crash when debugging large enclaves

### ISSUE THREAD

**Issue by anakrish** [intel/linux-sgx#444]
The debugger crashes when debugging large enclaves.

To reproduce, in SampleCode/SampleEnclave/Enclave/Enclave.config.xml change
TCSNum to 20
StackMaxSize to 0x3E8000 (which is 4K * 1000)

```bash
$ make clean
$ make
$ sgx-gdb --args ./app ./enclave.signed.so 
(gdb) r
...
/opt/intel/sgxsdk/bin/sgx-gdb: line 49: 47288 Killed                  LD_PRELOAD=$SGX_LIBRARY_PATH/libsgx_ptrace.so $GDB -iex "directory $GDB_SGX_PLUGIN_PATH" -iex "source $GDB_SGX_PLUGIN_PATH/gdb_sgx_plugin.py" -iex "set environment LD_PRELOAD" -iex "add-auto-load-safe-path /usr/lib" "$@"

```

This is the same issue as https://github.com/openenclave/openenclave/issues/2134

**Comment by anakrish** [comment]
To reproduce with plain GDB, make the changes to Enclave.config.xml as describe above.

```bash
$ make clean
$ make
$ gdb --args ./app ./enclave.signed.so 
(gdb) b generate_enclave_debug_event   
Function "generate_enclave_debug_event" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (generate_enclave_debug_event) pending.

(gdb) r
...
Breakpoint 1, generate_enclave_debug_event (code=2711625409, enclave_info=0x55555576f8d8) at debugger_support.cpp:92
92	    if(URTS_EXCEPTION_POSTINITENCLAVE == code)

(gdb) p/x *(uint64_t*)  enclave_info->tcs_list->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info.TCS_address  

Killed


```

**Comment by xxu36** [comment]
Hi, in my local environment, I cannot reproduce this issue. For the 1st case you mentioned above, the debugger ran successfully to the end. For the 2nd cast, I followed your steps and got the following result:
(gdb) p/x *(uint64_t*)  enclave_info->tcs_list->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info->next_tcs_info.TCS_address
Cannot access memory at address 0x0

'Cannot access memory at address 0x0' is reasonable here because when the gdb first stops here, some next_tcs_info pointers is NULL. However, no crash is observed.

I ran the test on a Ubuntu 1804 machine with gcc 7.4.0 and gdb 8.1.0 SGX psw and sdk code are the latest on master branch. The driver code I use comes from https://github.com/intel/linux-sgx-driver at branch sgx2 which supports EDMM feature.

Could you provide some information about your testing platform? Thanks

**Comment by anakrish** [comment]
@xxu36 Can you confirm if you are indeed setting the TCSNum to 20 in the config file?
next_tcs_info ought not to be null since there are 20 tcs and we are trying to access the 11th one.

If it is still reproducible, can you try setting the number of tcs to a very large value say 100?

**Comment by anakrish** [comment]
@xxu36 I am running the reproduction steps on an ACC VM. Information about the kernel as well as the sgx driver

```bash
$ uname -a
Linux akoevml 5.0.0-1018-azure #19~18.04.1-Ubuntu SMP Wed Aug 21 05:13:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

$ modinfo intel_sgx
filename:       /lib/modules/5.0.0-1018-azure/updates/dkms/intel_sgx.ko
license:        Dual BSD/GPL
author:         Jarkko Sakkinen <jarkko.sakkinen@linux.intel.com>
description:    Intel SGX Driver
srcversion:     0253BC73E4B1968CF600DAD
alias:          acpi*:INT0E0C:*
depends:        
retpoline:      Y
name:           intel_sgx
vermagic:       5.0.0-1018-azure SMP mod_unload 
signat:         PKCS#7
signer:         
sig_key:        
sig_hashalgo:   md4
```

The SDK and PSW are built from the lastest master branch.

**Comment by xxu36** [comment]
@anakrish Yes, I set the TCSNum to 20, but the result I mentioned above ran on a EDMM enabled driver, so not all the TCS were created during loading time.

But I also ran the test using the  driver on the master branch, which has no EDMM feature. With this driver, I can get the TCS address, no crash is observed even. Even I configured the heap size, stack max size and tcs num to a larger number, still no crash.

I will check whether we have a ACC vm at hand and if so I'll do some tests using that environment.

**Comment by anakrish** [comment]
@xxu36 Thanks for trying it out. I will try to reproduce it on a non ACC VM.

**Comment by haitaohuang** [comment]
Following instruction at https://wiki.ubuntu.com/Kernel/LTSEnablementStack to install the HWE stack on a Ubtuntu 18.04 desktop running on GLK NUC (94M EPC). The crash is reproducible with DCAP driver.
